%% ----------------------------------------------------------------
%% Thesis.tex -- MAIN FILE (the one that you compile with LaTeX)
%% ---------------------------------------------------------------- 

% Set up the document
\documentclass[a4paper, 11pt, oneside]{Thesis}  % Use the "Thesis" style, based on the ECS Thesis style by Steve Gunn
\graphicspath{Figures/}  % Location of the graphics files (set up for graphics to be in PDF format)

% Include any extra LaTeX packages required
\usepackage[round, comma]{natbib}  % Use the "Natbib" style for the references in the Bibliography
\usepackage{verbatim}  % Needed for the "comment" environment to make LaTeX comments
\usepackage{vector}  % Allows "\bvec{}" and "\buvec{}" for "blackboard" style bold vectors in maths
\usepackage{comment}
\usepackage{xcolor}
\usepackage{soul}
%%%% commands
\newcommand{\note}[1]{\textcolor{red}{#1}}
%%%%


\bibliographystyle{plainnat}

\hypersetup{urlcolor=blue, colorlinks=true}  % Colours hyperlinks in blue, but this can be distracting if there are many links.

%% ----------------------------------------------------------------
\begin{document}
\frontmatter      % Begin Roman style (i, ii, iii, iv...) page numbering

% Set up the Title Page
\title  {Thesis Title}
\authors  {\texorpdfstring
            {\href{http://www.hiraksarkar.com}{Hirak Sarkar}}
            {Hirak Sarkar}
            }
%\addresses  {\groupname\\\deptname\\\univname}  % Do not change this here, instead these must be set in the "Thesis.cls" file, please look through it instead
\date       {\today}
\subject    {}
\keywords   {}

\maketitle
%% ----------------------------------------------------------------

\setstretch{1.3}  % It is better to have smaller font and larger line spacing than the other way round

% Define the page headers using the FancyHdr package and set up for one-sided printing
%\fancyhead{}  % Clears all page headers and footers
%\rhead{\thepage}  % Sets the right side header to show the page number
%\lhead{}  % Clears the left side page header
\pagestyle{fancy}
%\renewcommand{\subsubsectionmark}[1]{\markright{\thesubsubsection\ #1}}
\fancyhf{}
%\rhead{\fancyplain{}{$<$Name of the report$>$}} % predefined ()
%\lhead{\fancyplain{}{\rightmark }}
%\fancyhead[L]{\emph{\rightmark}}
%\fancyhead[R]{\thepage}
  % Finally, use the "fancy" page style to implement the FancyHdr headers

%% ----------------------------------------------------------------
% Declaration Page required for the Thesis, your institution may give you a different text to place here
\begin{comment}
\Declaration{

\addtocontents{toc}{\vspace{1em}}  % Add a gap in the Contents, for aesthetics

I, AUTHOR NAME, declare that this thesis titled, `THESIS TITLE' and the work presented in it are my own. I confirm that:

\begin{itemize} 
\item[\tiny{$\blacksquare$}] This work was done wholly or mainly while in candidature for a research degree at this University.
 
\item[\tiny{$\blacksquare$}] Where any part of this thesis has previously been submitted for a degree or any other qualification at this University or any other institution, this has been clearly stated.
 
\item[\tiny{$\blacksquare$}] Where I have consulted the published work of others, this is always clearly attributed.
 
\item[\tiny{$\blacksquare$}] Where I have quoted from the work of others, the source is always given. With the exception of such quotations, this thesis is entirely my own work.
 
\item[\tiny{$\blacksquare$}] I have acknowledged all main sources of help.
 
\item[\tiny{$\blacksquare$}] Where the thesis is based on work done by myself jointly with others, I have made clear exactly what was done by others and what I have contributed myself.
\\
\end{itemize}
 
 
Signed:\\
\rule[1em]{25em}{0.5pt}  % This prints a line for the signature
 
Date:\\
\rule[1em]{25em}{0.5pt}  % This prints a line to write the date
}
\clearpage  % Declaration ended, now start a new page
\end{comment}
%% ----------------------------------------------------------------
% The "Funny Quote Page"
\pagestyle{empty}  % No headers or footers for the following pages

\null\vfill
% Now comes the "Funny Quote", written in italics
\textit{``Nature uses only the longest threads to weave her patterns, so that each small piece of her fabric reveals the organization of the entire tapestry.''}

\begin{flushright}
— Richard P. Feynman
\end{flushright}

\vfill\vfill\vfill\vfill\vfill\vfill\null
\clearpage  % Funny Quote page ended, start a new page
%% ----------------------------------------------------------------

% The Abstract Page
\addtotoc{Abstract}  % Add the "Abstract" page entry to the Contents
\abstract{
\addtocontents{toc}{\vspace{1em}}  % Add a gap in the Contents, for aesthetics

Recent technological advances in high-throughput sequencing have posed a new challenge of designing robust and efficient algorithms for downstream analysis. This rapid growth has also widened the scope for application of more sophisticated statistical techniques to discover biological insights. In this report, we have focused on analysis of biological sequencing data derived from RNA-seq, an experimental protocol for assaying the entire transcriptome ( i.e., the collection of expressed transcripts) of a given sample. RNA-seq is tremendously useful for a wide variety of tasks ranging from differential expression analysis to the discovery of previously unannotated genes and transcripts. In this work, we used the initial mapping results from \qm, a state-of-the-art mapping algorithm, to construct a more compact representation, termed as {\it equivalence classes} of the RNA-seq experiment. We have shown the usefulness of such representation via two different applications: \rapclust and \quark..

Our first tool, \rapclust is designed to assist in the analysis of \denovo transcriptomes. \rapclust introduces a novel and highly efficient method for clustering contigs from \denovo transcriptome assemblies, exploiting the multi-mapping sequence fragments as evidence of contigs that may derive from the same underlying gene. Specifically, we cast the problem of clustering contigs as the problem of clustering a sparse graph. Compared to other state-of-the-art tools, \rapclust is substantially faster.  Moreover, \rapclust produces clusters that with a closer correspondence to the true genes of origin than other tools.  The methodology of \rapclust also yields accurate estimates of expression for the resulting clusters (i.e., genes). Our second tool, \quark, uses the concept of {\it equivalence classes} to compress RNA-seq read data. \quark provides state-of-the-art compression rates, and introduces the concept of ``semi-reference-based" compression, in which a reference is assumed at the compressor, but not required at the decompressor.


%Recent technological advances in high throughput sequencing has posed a new challenge of designing robust and efficient algorithms for analysis. This rapid growth has also widen the scope for application of more sophisticated statistical techniques to discover biological insights. In this report we have focused on analysis of biological sequencing data derived from RNA-seq. RNA-seq is an experimental protocol for assaying the entire transcriptome ( i.e. the collection of expressed transcripts) of a given sample. RNA-seq is tremendously useful for various range of tasks from differential expression analysis to the discovery of previously unannotated genes and transcripts. We use the initial mapping results from \qm: a state of the art mapping algorithm, to construct a more compact representation, termed as {\it equivalence classes} of the RNA-seq experiment. We have shown the usefulness such representation via two different applications.

%Our first tool, \rapclust is designed to assist in the analysis of \denovo transcriptomes. \rapclust introduces a novel and highly-efficient method for clustering contigs from de novo transcriptome assemblies, exploiting the multi-mapping sequence fragments as evidence of contigs that may derive from the same underlying gene. To be more specific we cast the problem of clustering contigs as the problem of clustering a sparse graph. Compared to other state of the art tools, \rapclust is tremendously faster.  Moreover, \rapclust produces clusters that more closely correspond to the true genes of origin than other tools.  The methodology of \rapclust also yields accurate estimates of expression for the resulting clusters (i.e. genes). Our second tool \quark uses the concept of {\it equivalence classes} to compress RNA-seq read data. \quark provides state-of-the-art compression rates, and introduces the concept of ``semi-reference-based" compression, in which a reference is assumed at the compressor, but not required at the decompressor.

%My research focuses on the design of robust and efficient algorithms for analyzing biological data, with the goal of obtaining biological insights from typically large datasets. The major concentration of my study has been the analysis of biological sequencing data — specifically, RNA-seq. RNA-seq is an experimental protocol for assaying the entire transcriptome ( i.e. the collection of expressed transcripts) of a given sample. RNA-seq is tremendously useful for various range of tasks from differential expression analysis to the discovery of previously unannotated genes and transcripts. 

%A crucial application of RNA-seq is de-novo transcriptome analysis, in which we wish to study the transcriptomes of organisms for which no genomic assembly is available.  We have developed a tool, RapClust, to assist in the analysis of de novo transcriptomes. RapClust introduces a novel and highly-efficient method for clustering contigs from de novo transcriptome assemblies, exploiting the multi-mapping sequence fragments as evidence of contigs that may derive from the same underlying gene. To be more specific we cast the problem of clustering contigs as the problem of clustering a sparse graph. Compared to other state of the art tools, RapClust is tremendously faster.  Moreover, RapClust produces clusters that more closely correspond to the true genes of origin than other tools.  The methodology of RapClust also yields accurate estimates of expression for the resulting clusters (i.e. genes). 

%I will also discuss another of our tools, Quark (yet to release), which adopts a novel methodology  to compress RNA-seq read data. Quark provides state-of-the-art compression rates, and introduces the concept of “semi-reference-based” compression, in which a reference is assumed at the compressor, but not required at the decompressor.  Quark generates a very compact representation of the read sequences, that can be efficiently used for downstream analysis.


}

\clearpage  % Abstract ended, start a new page
%% ----------------------------------------------------------------

\setstretch{1.3}  % Reset the line-spacing to 1.3 for body text (if it has changed)

% The Acknowledgements page, for thanking everyone

\acknowledgements{
\addtocontents{toc}{\vspace{1em}}  % Add a gap in the Contents, for aesthetics

I would like to thank my advisor Prof. Rob Patro for his patience and continuous support throughout the project. My fellow students Avi, Laraib, Nitish, Komal and Mohsin for patiently listening and commenting on the progress of the project in every group meeting.
}
\clearpage  % End of the Acknowledgements
%% ----------------------------------------------------------------

\pagestyle{fancy}  %The page style headers have been "empty" all this time, now use the "fancy" headers as defined before to bring them back


%% ----------------------------------------------------------------
\lhead{\emph{Contents}}  % Set the left side page header to "Contents"
\tableofcontents  % Write out the Table of Contents

%% ----------------------------------------------------------------
\lhead{\emph{List of Figures}}  % Set the left side page header to "List if Figures"
\listoffigures  % Write out the List of Figures

%% ----------------------------------------------------------------
\lhead{\emph{List of Tables}}  % Set the left side page header to "List of Tables"
\listoftables  % Write out the List of Tables

%% ----------------------------------------------------------------
\setstretch{1.5}  % Set the line spacing to 1.5, this makes the following tables easier to read
\clearpage  % Start a new page
%\lhead{\emph{Abbreviations}}  % Set the left side page header to "Abbreviations"
%\listofsymbols{ll}  % Include a list of Abbreviations (a table of two columns)
%{
% \textbf{Acronym} & \textbf{W}hat (it) \textbf{S}tands \textbf{F}or \\
%\textbf{LAH} & \textbf{L}ist \textbf{A}bbreviations \textbf{H}ere \\

%}

%% ----------------------------------------------------------------
%\clearpage  % Start a new page
%\lhead{\emph{Physical Constants}}  % Set the left side page header to "Physical Constants"
%\listofconstants{lrcl}  % Include a list of Physical Constants (a four column table)
%{
% Constant Name & Symbol & = & Constant Value (with units) \\
%Speed of Light & $c$ & $=$ & $2.997\ 924\ 58\times10^{8}\ \mbox{ms}^{-\mbox{s}}$ (exact)\\}

%% ----------------------------------------------------------------
% \clearpage  %Start a new page
% \lhead{\emph{Symbols}}  % Set the left side page header to "Symbols"
% \listofnomenclature{lll}  % Include a list of Symbols (a three column table)
% {
% % symbol & name & unit \\
% $a$ & distance & m \\
% $P$ & power & W (Js$^{-1}$) \\
% & & \\ % Gap to separate the Roman symbols from the Greek
% $\omega$ & angular frequency & rads$^{-1}$ \\
% }
% %% ----------------------------------------------------------------
% % End of the pre-able, contents and lists of things
% % Begin the Dedication page

% \setstretch{1.3}  % Return the line spacing back to 1.3

% \pagestyle{empty}  % Page style needs to be empty for this page
% \dedicatory{For/Dedicated to/To my\ldots}

\addtocontents{toc}{\vspace{2em}}  % Add a gap in the Contents, for aesthetics


%% ----------------------------------------------------------------
\mainmatter	  % Begin normal, numeric (1,2,3...) page numbering
\pagestyle{fancy}  % Return the page headers back to the "fancy" style

% Include the chapters of the thesis, as separate files
% Just uncomment the lines as you write the chapters



%\lhead{\emph{\thesection}} 
\input{Chapters/Chapter1} % Introduction

\input{Chapters/Chapter2} % Background Theory 

\input{Chapters/Chapter3} % Experimental Setup

\input{Chapters/Chapter4} % Experiment 1

%\input{Chapters/Chapter5} % Experiment 2

%\input{Chapters/Chapter6} % Results and Discussion

%\input{Chapters/Chapter7} % Conclusion

%% ----------------------------------------------------------------
% Now begin the Appendices, including them as separate files

\addtocontents{toc}{\vspace{2em}} % Add a gap in the Contents, for aesthetics

%\appendix % Cue to tell LaTeX that the following 'chapters' are Appendices

%\input{Appendices/AppendixA}	% Appendix Title

%\input{Appendices/AppendixB} % Appendix Title

%\input{Appendices/AppendixC} % Appendix Title

\addtocontents{toc}{\vspace{2em}}  % Add a gap in the Contents, for aesthetics
\backmatter

%% ----------------------------------------------------------------
\label{Bibliography}
\lhead{\emph{Bibliography}}  % Change the left side page header to "Bibliography"
\bibliographystyle{unsrtnat}  % Use the "unsrtnat" BibTeX style for formatting the Bibliography
\bibliography{Bibliography}  % The references (bibliography) information are stored in the file named "Bibliography.bib"

\end{document}  % The End
%% ----------------------------------------------------------------